src/pkg/user/service/service.go

```diff
@@ -22,77 +28,247 @@ type Service struct {
 	Proc  *processing.Proc
 }
 
+type TokenMetadata struct {
+	UUID   string
+	UserID string
+}
+
+type TokenDetails struct {
+	AccessToken  string
+	RefreshToken string
+	AccessUUID   string
+	RefreshUUID  string
+	AtExpires    int64
+	RtExpires    int64
+}
+
 const (
-	N            = 1048576
-	r            = 8
-	p            = 1
-	keyLen       = 32
-	saltLenBytes = 16
+	N      = 1048576
+	r      = 8
+	p      = 1
+	keyLen = 32
 )
 
 func NewUserService(cfg *config.Config, infra *infrastructure.Infra) *Service {
-	proc := processing.New(
+	process := processing.New(
 		infra.LocalDB,
 		infra.EthClient,
 		infra.FilecoinClient,
 		infra.IpfsClient,
 		cfg.Storage.Localfile.Path,
 	)
 
-	proc.Start()
+	process.Start()
 
 	return &Service{
 		Infra: infra,
-		Proc:  proc,
+		Proc:  process,
 	}
 }
 
 func (s *Service) Register(ctx context.Context, procRequest *proc.Request, user *model.UserCreateRequest) (err error) {
-	_, userPrivateKey, err := s.Infra.Keystore.Get(user.UserID)
+	ehrSystemID := ctx.(*gin.Context).GetString("ehrSystemID")
+	address, pwdHash, err := s.getUserAddressAndHash(ehrSystemID, user.UserID, user.Password)
+
+	if err != nil {
+		return err
+	}
+
+	requestID := ctx.(*gin.Context).GetString("reqId")
+
+	txHash, err := s.Infra.Index.UserAdd(requestID, address, user.UserID, user.Role, pwdHash)
 	if err != nil {
-		return fmt.Errorf("Keystore.Get error: %w userID %s", err, user.UserID)
+		return fmt.Errorf("Index.UserAdd error: %w", err)
+	}
+
+	procRequest.AddEthereumTx(proc.TxDeleteDoc, txHash)
+
+	return nil
+}
+
+func (s *Service) Login(ctx context.Context, user *model.UserAuthRequest) (err error) {
+	ehrSystemID := ctx.(*gin.Context).GetString("ehrSystemID")
+	address, pwdHash, err := s.getUserAddressAndHash(ehrSystemID, user.UserID, user.Password)
+
+	if err != nil {
+		return err
+	}
+
+	userHash, err := s.Infra.Index.GetUserPasswordHash(ctx, address)
+	if err != nil {
+		return fmt.Errorf("Login.GetUserPasswordHash error: %w", err)
+	}
+
+	if !bytes.Equal(pwdHash, userHash) {
+		return errors.ErrFieldIsIncorrect("Password")
+	}
+
+	return nil
+}
+
+func (s *Service) getUserAddressAndHash(ehrSystemID, userID, password string) (eth_common.Address, []byte, error) {
+	_, userPrivateKey, err := s.Infra.Keystore.Get(userID)
+	if err != nil {
+		return eth_common.Address{}, nil, fmt.Errorf("Keystore.Get error: %w userID %s", err, userID)
 	}
 
 	privateUserKey := userPrivateKey[:]
 
 	privateKey, err := crypto.ToECDSA(privateUserKey)
 	if err != nil {
-		return fmt.Errorf("crypto.ToECDSA error: %w userID %s", err, user.UserID)
+		return eth_common.Address{}, nil, fmt.Errorf("crypto.ToECDSA error: %w userID %s", err, userID)
 	}
 
 	address := crypto.PubkeyToAddress(privateKey.PublicKey)
-	ehrSystemID := ctx.(*gin.Context).GetString("ehrSystemID")
 
-	pwdHash, err := s.generateHash(user.UserID, ehrSystemID, user.Password)
+	pwdHash, err := s.generateHash(privateUserKey, userID, ehrSystemID, password)
```

Comment:
Bad idea, salt must be a random but never a privateKey!!!

Reply:
\+